### PR TITLE
IPC: fix component new fail issue.

### PR DESCRIPTION
**Pull request by keyonjie**
We need to new component with the whole ipc struction passed in
ipc_comp_new(), e.g. struct sof_ipc_comp_dai, but not only the struct
sof_ipc_comp, otherwise, we will fail to new components and see errors
in log such as:
dai.c:208 	dai_new() error: dai_get() failed to create DAI. type:0,
index:-1106776464

Here change to pass in the whole _ipc->comp_data to fix this kind of
issues.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

**Comment by keyonjie** [comment]
> @keyonjie
> Nice catch.
> So this bug only happen with XCC binary?
> Did not see such issue with GCC binaries.

it's strange to me why it doesn't happen with GCC, it should happen both.

File: src/include/sof/ipc.h [hunk 1]
```diff
@@ -96,7 +96,7 @@ struct dai_config;
 			trace_ipc("ipc: hdr 0x%x rx (%d) > tx (%d)",	\
 				  rx->cmd, rx_size, tx->size);		\
 		} else if (tx->size > rx_size) {			\
-			memcpy(rx, tx, tx->size);			\
+			memcpy(rx, tx, rx_size);			\
```
Comment (ranj063) [inline src/include/sof/ipc.h]: @keyonjie should this be in a separate patch?

Comment (lgirdwood) [reply src/include/sof/ipc.h]: It's ok, I'll merge in one PR.

**Comment by lgirdwood** [comment]
> > @keyonjie
> > Nice catch.
> > So this bug only happen with XCC binary?
> > Did not see such issue with GCC binaries.
> 
> it's strange to me why it doesn't happen with GCC, it should happen both.

Yep, quite strange. I do know that some layout of some data structures is different between compilers....

**Comment by jajanusz** [comment]
> > > @keyonjie
> > > Nice catch.
> > > So this bug only happen with XCC binary?
> > > Did not see such issue with GCC binaries.
> > 
> > 
> > it's strange to me why it doesn't happen with GCC, it should happen both.
> 
> Yep, quite strange. I do know that some layout of some data structures is different between compilers....

@keyonjie @lgirdwood most likely caused by stack reordering - there is no flag to stop it in xcc, but if are curious you can try it by wrapping all local variables in struct - this should stop all compilers from reordering them
